pgxn/neon/pagestore_smgr.c

```diff
@@ -683,55 +682,40 @@ prefetch_set_unused(uint64 ring_index)
 		compact_prefetch_buffers();
 }
 
+/*
+ * Send one prefetch request to the pageserver. To wait for the response, call
+ * prefetch_wait_for().
+ */
 static void
-prefetch_do_request(PrefetchRequest *slot, bool *force_latest, XLogRecPtr *force_lsn)
+prefetch_do_request(PrefetchRequest *slot, XLogRecPtr *force_request_lsn, XLogRecPtr *force_not_modified_since)
 {
 	bool		found;
 	NeonGetPageRequest request = {
-		.req.tag = T_NeonGetPageRequest,
-		.req.latest = false,
-		.req.lsn = 0,
+		.req.tag = T_NeonGetPageV2Request,
+		/* lsn and not_modified_since are filled in below */
 		.rinfo = BufTagGetNRelFileInfo(slot->buftag),
 		.forknum = slot->buftag.forkNum,
 		.blkno = slot->buftag.blockNum,
 	};
 
-	if (force_lsn && force_latest)
+	Assert(((force_request_lsn != NULL)  == (force_not_modified_since != NULL)));
+
+	if (force_request_lsn)
 	{
-		request.req.lsn = *force_lsn;
-		request.req.latest = *force_latest;
-		slot->actual_request_lsn = slot->effective_request_lsn = *force_lsn;
+		request.req.lsn = *force_request_lsn;
+		request.req.not_modified_since = *force_not_modified_since;
+		slot->request_lsn = *force_request_lsn;
+		slot->not_modified_since = *force_not_modified_since;
 	}
 	else
 	{
-		XLogRecPtr	lsn = neon_get_request_lsn(
-											   &request.req.latest,
-											   BufTagGetNRelFileInfo(slot->buftag),
-											   slot->buftag.forkNum,
-											   slot->buftag.blockNum
-			);
-
-		/*
-		 * Note: effective_request_lsn is potentially higher than the
-		 * requested LSN, but still correct:
-		 *
-		 * We know there are no changes between the actual requested LSN and
-		 * the value of effective_request_lsn: If there were, the page would
-		 * have been in cache and evicted between those LSN values, which then
-		 * would have had to result in a larger request LSN for this page.
-		 *
-		 * It is possible that a concurrent backend loads the page, modifies
-		 * it and then evicts it again, but the LSN of that eviction cannot be
-		 * smaller than the current WAL insert/redo pointer, which is already
-		 * larger than this prefetch_lsn. So in any case, that would
-		 * invalidate this cache.
-		 *
-		 * The best LSN to use for effective_request_lsn would be
-		 * XLogCtl->Insert.RedoRecPtr, but that's expensive to access.
-		 */
-		slot->actual_request_lsn = request.req.lsn = lsn;
-		prefetch_lsn = Max(prefetch_lsn, lsn);
```

Comment:
It seems to be still reasonable to maintain `prefetch_lsn`. May be it is better to rename it to `max_prefetch_lsn` or whatever else. The motivation is the following if we have sent prefetch request with `not_modified_since=LSN1`, then PS in any case has to wait until `LSN1` is applied. So there is no sense to send next requests with smaller  `not_modified_since` And what is the advantage of sending prefetch request with `not_modified_since` larger than `last-modified-lsn` we know at the moment of issuing this request? It reduce probability of expired (deteriorated) prefetch request. If out `last-written-lsn` cache is precise, there will be no such problem. But it is not the case now:
1. `last-written-lsn` cache is limited in size and uses eviction algorithm
2. after compute restart this cache is empty.
So this is actually quite important optimization affecting prefetch performance.

Reply:
> It seems to be still reasonable to maintain `prefetch_lsn`. May be it is better to rename it to `max_prefetch_lsn` or whatever else. The motivation is the following if we have sent prefetch request with `not_modified_since=LSN1`, then PS in any case has to wait until `LSN1` is applied. So there is no sense to send next requests with smaller `not_modified_since`

Right, the pageserver currently only uses the 'not_modified_since' value to decide whether to wait or not. However, it could use it for further optimizations: it might be able to skip over recent delta layers if it knows that they cannot contain the page being requested. So if we can give it more accurate information, it would be nice to do so.

> And what is the advantage of sending prefetch request with `not_modified_since` larger than `last-modified-lsn` we know at the moment of issuing this request? It reduce probability of expired (deteriorated) prefetch request. If out `last-written-lsn` cache is precise, there will be no such problem. But it is not the case now:
> 
>     1. `last-written-lsn` cache is limited in size and uses eviction algorithm
> 
>     2. after compute restart this cache is empty.
>        So this is actually quite important optimization affecting prefetch performance.

Hmm, I don't quite understand that.

Imagine that there is a request in the prefetch queue with LSN range [100, 500]. If we now get a new smgrread call for the same page, and calculate the range for the new call to be [400, 600], the old request can be used because the ranges overlap. If the old request had used a more recent `not_modified_since` hint, so that it was for example [200,500] or [450, 500], it would not change anything.

Reply:
> So if we can give it more accurate information, it would be nice to do so.

I agree with you, but there is not way now to efficiently skip layers: just LSN&key range us not enough and reading layer's Btree is mostly the same as reading layers.

> Hmm, I don't quite understand that.

Sorry, I have to agree: it is not currently needed - range overlap check is eliminating need in it.

Reply:
Ok, great!

Reply:
@MMeent can you also review these prefetching changes, please?

Reply:
This logic is a very critical for correctness, and if we get it wrong, the symptoms could be very weird, e.g. a slightly out-of-date page version might reappear. Because of that, I moved it to a separate function, mainly to have place for a comment to explain the logic. This is very little code, but seems worth a long comment to explain the logic: @arssher and I were confused about this as it was in the original PR #6718, and you @knizhnik had that question about the overlapping ranges in this PR too.